### ISSUE TITLE
[rustdoc] Feature Request: Allow code wrapping

### ISSUE THREAD

**Issue by aramrw** [rust-lang/rust#127334]
Would be nice to have a setting (or just a button) that could have the code wrap, gets tiring if you are reading lots of code on your phone to scroll back and forth.

My bad if this has been requested before, I didn't see it in any issues yet

**Comment by GuillaumeGomez** [comment]
In case it's not clear for everyone, let me expand a bit (please correct me if I'm wrong @aramrw):

Currently, the code examples in Rust HTML documentation are scrollable if they are too wide. The idea here would be to add a setting allowing for the code examples to instead wrap to the next line.

Now for the implementation: we need to be careful how this new setting would interact with the line numbers on the left (if the setting is enabled).

One thing I'm not clear about: @aramrw, would you want this setting to be applied to both code examples and source code pages or only to code examples?

I think it's fine to add such a setting. If @rust-lang/rustdoc-frontend agrees, let's add it.

**Comment by aramrw** [comment]
> One thing I'm not clear about: @aramrw, would you want this setting to be applied to both code examples and source code pages or only to code examples?
> 

If it's possible to add it for both the code examples & the source code it would be much appreciated 👍🏼

**Comment by GuillaumeGomez** [comment]
A table has a limit of 10.000 rows, so might be problematic on source code page with some very big files.

**Comment by aramrw** [comment]
> A table has a limit of 10.000 rows, so might be problematic on source code page with some very big files.

If that's the case then only wrapping code blocks is fine because _(on iphone at least)_, the source code allows you to pinch in and out very easily to see all the text, or at least most of it.

However the code block examples don't have that, so to see all the text you need to go into the browsers menu and keep messing with the zoom in & out buttons.

**Comment by jsha** [comment]
https://github.com/rust-lang/rust/issues/88545

Looks like it was grid layout and 1000 items, maybe?

**Comment by camelid** [comment]
One idea: instead of having an option for this, maybe just make wrapping the default on mobile devices (i.e., screens that are sufficiently narrow by some heuristic). I'm wary of requiring too much configuration of rustdoc.

**Comment by GuillaumeGomez** [comment]
> #88545 Looks like it was grid layout and 1000 items, maybe?

I was looking for this discussion, thanks so much! So it was not for tables but for grids, good to know! One less problem to worry about.

> One idea: instead of having an option for this, maybe just make wrapping the default on mobile devices (i.e., screens that are sufficiently narrow by some heuristic). I'm wary of requiring too much configuration of rustdoc.

We enter in the tricky area of personal taste. It's a case where being able to ask all rust doc users would be super useful.

**Comment by jsha** [comment]
I've been meaning to come here and post exactly what @camelid said. Our default position should be to not introduce new settings but try to just do the best thing. If we put it behind a setting, very few people will actually know about it and use it, but we will still bear the full burden of maintaining both versions.

In this case, the right thing is clearly to line-wrap on all platforms (not just mobile). Horizontal scrolling is a terrible experience on desktop / wide screens too.

[This page has an example of doing line numbers with wrapping](https://joriszwart.nl/articles/css-line-numbers). It uses CSS counters, and seems pretty tidy.

**Comment by GuillaumeGomez** [comment]
It can work for code examples, but not for source code as we need them to be link. But noted for the horizontal scrolling. Will make our work simpler.

**Comment by GuillaumeGomez** [comment]
Another issue I didn't think of is that we use style with CSS classes across lines. It would greatly increase the size of the DOM. This feature is definitely more tricky than I initially thought...

**Comment by aramrw** [comment]
any updates...?

**Comment by GuillaumeGomez** [comment]
Yep, found an acceptable middle-ground for this feature to be done in https://github.com/rust-lang/rust/pull/136829. It's just the first step but adding this setting should be pretty easy/quick from there.

**Comment by GuillaumeGomez** [comment]
Opened PR to add the feature in https://github.com/rust-lang/rust/pull/136991.
